test/integration/pubsub_msg_seen_cache_test.go

```diff
@@ -0,0 +1,244 @@
+					t.Logf("sending [%s] with duplicate message ID at [%s]", msg, now)
+				} else {
+					msgID = pubsub.DefaultMsgIdFn(pmsg)
+					t.Logf("sending [%s] with unique message ID at [%s]", msg, now)
+				}
+				return msgID
+			}),
+		)
+		return append(
+			info.FXOptions,
+			fx.Provide(libp2p2.TopicDiscovery()),
+			fx.Decorate(libp2p2.GossipSub(pubsubOptions...)),
+		), nil
+	})
+
+	producerNode, err = mockNode(ctx, mn, false, "") // PubSub configuration comes from overrides above
+	if err != nil {
+		t.Fatal(err)
+	}
+	producerPeerID = producerNode.PeerHost.ID()
+	defer producerNode.Close()
+
+	t.Logf("bootstrap peer=%s, consumer peer=%s, producer peer=%s", bootstrapPeerID, consumerPeerID, producerPeerID)
+
+	producerAPI, err := coreapi.NewCoreAPI(producerNode)
+	if err != nil {
+		t.Fatal(err)
+	}
+	consumerAPI, err := coreapi.NewCoreAPI(consumerNode)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = mn.LinkAll()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	bis := bootstrapNode.Peerstore.PeerInfo(bootstrapNode.PeerHost.ID())
+	bcfg := bootstrap.BootstrapConfigWithPeers([]peer.AddrInfo{bis})
+	if err = producerNode.Bootstrap(bcfg); err != nil {
+		t.Fatal(err)
+	}
+	if err = consumerNode.Bootstrap(bcfg); err != nil {
+		t.Fatal(err)
+	}
+
+	// Set up the consumer subscription
+	const TopicName = "topic"
+	consumerSubscription, err := consumerAPI.PubSub().Subscribe(ctx, TopicName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Utility functions defined inline to include context in closure
+	now := func() string {
+		return time.Now().Format(time.StampMilli)
+	}
+	ctr := 0
+	msgGen := func() string {
+		ctr++
+		return fmt.Sprintf("msg_%d", ctr)
+	}
+	produceMessage := func() string {
+		msgTxt := msgGen()
+		err = producerAPI.PubSub().Publish(ctx, TopicName, []byte(msgTxt))
+		if err != nil {
+			t.Fatal(err)
+		}
+		return msgTxt
+	}
+	consumeMessage := func(msgTxt string, shouldFind bool) {
+		// Set up a separate timed context for receiving messages
+		rxCtx, rxCancel := context.WithTimeout(context.Background(), time.Second)
+		defer rxCancel()
+		msg, err := consumerSubscription.Next(rxCtx)
+		if shouldFind {
+			if err != nil {
+				t.Logf("did not receive [%s] by [%s]", msgTxt, now())
+				t.Fatal(err)
+			}
+			t.Logf("received [%s] at [%s]", string(msg.Data()), now())
+			if !bytes.Equal(msg.Data(), []byte(msgTxt)) {
+				t.Fatalf("consumed data [%s] does not match published data [%s]", string(msg.Data()), msgTxt)
+			}
+		} else {
+			if err == nil {
+				t.Logf("received [%s] at [%s]", string(msg.Data()), now())
+				t.Fail()
+			}
+			t.Logf("did not receive [%s] by [%s]", msgTxt, now())
+		}
+	}
+
+	// Send message 1 with the message ID we're going to duplicate later
+	sendDupMsg = true
+	msgTxt := produceMessage()
+	consumeMessage(msgTxt, true) // should find message
+
+	// Send message 2 with the same message ID as before
+	sendDupMsg = true
+	msgTxt = produceMessage()
+	consumeMessage(msgTxt, false) // should NOT find message
```

Comment:
```suggestion
	consumeMessage(msgTxt, false) // should NOT find message, because it got deduplicated (sent twice within the SeenMessagesTTL window)
```